irobot_events_executor/src/rclcpp/executors/events_executor/events_executor.cpp

```diff
@@ -52,7 +52,22 @@ EventsExecutor::EventsExecutor(
 
 EventsExecutor::~EventsExecutor()
 {
-  spinning.store(false);
+  // Set 'spinning' to false and get previous value
+  auto executor_was_spinning = spinning.exchange(false);
+
+  if (executor_was_spinning) {
+    // Trigger the shutdown guard condition.
+    // This way, the 'events_queue_' will wake up since a new event has arrived.
+    // Then since 'spinning' is now false, the spin loop will exit.
+    shutdown_guard_condition_->trigger();
+
+    // The timers manager thread is stopped at the end of spin().
+    // We have to wait for timers manager thread to exit, otherwise
+    // the 'timers_manager_' will be destroyed while still being used on spin().
+    while (timers_manager_->is_running()) {
+      std::this_thread::sleep_for(1ms);
```

Comment:
This could be replaced by a `TimersManager::wait_until_stop()` which waits on the atomic bool `running_` to change to false 
https://en.cppreference.com/w/cpp/atomic/atomic/wait

Reply:
I'm not familiar with this method, but it looks like a good solution

Reply:
Oh nevermind, this method is available since C++20.